src/Module/BaseApi.php

```diff
@@ -195,80 +187,108 @@ public static function getRequest(array $defaults)
 	 *               'authenticated' => return status,
 	 *               'user_record' => return authenticated user record
 	 */
-	protected static function login(string $scope)
+	protected static function login(string $scope): bool
 	{
-		if (empty(self::$current_user_id)) {
-			self::$current_token = self::getTokenByBearer();
-			if (!empty(self::$current_token['uid'])) {
-				self::$current_user_id = self::$current_token['uid'];
-			} else {
-				self::$current_user_id = 0;
-			}
+		if (!self::getCachedCurrentUserIDFromRequest()) {
+			Logger::debug(API_LOG_PREFIX . 'failed', ['module' => 'api', 'action' => 'login', 'parameters' => $_SERVER]);
+			header('WWW-Authenticate: Basic realm="Friendica"');
+			throw new HTTPException\UnauthorizedException('This API requires login');
 		}
 
-		if (!empty($scope) && !empty(self::$current_token)) {
-			if (empty(self::$current_token[$scope])) {
-				Logger::warning('The requested scope is not allowed', ['scope' => $scope, 'application' => self::$current_token]);
-				DI::mstdnError()->Forbidden();
-			}
-		}
-
-		if (empty(self::$current_user_id)) {
-			// The execution stops here if no one is logged in
-			api_login(DI::app());
+		$current_token = self::getCachedTokenByBearer();
+		if ($current_token && empty($current_token[$scope])) {
+			Logger::warning('The requested scope is not allowed', ['scope' => $scope, 'application' => self::getCachedTokenByBearer()]);
+			DI::mstdnError()->Forbidden();
 		}
 
-		self::$current_user_id = api_user();
-
-		return (bool)self::$current_user_id;
+		return (bool)api_user();
 	}
 
 	/**
-	 * Get current application
+	 * Returns the current logged user id from the local variable or attempts to retrieve it from the request.
 	 *
-	 * @return array token
+	 * @return int
+	 * @throws HTTPException\ForbiddenException
+	 * @throws HTTPException\InternalServerErrorException
+	 * @throws HTTPException\UnauthorizedException
 	 */
-	protected static function getCurrentApplication()
+	public static function getCachedCurrentUserIdFromRequest(): int
```

Comment:
I must confess: Especially because of the lengths of the function name I will always assign the return value of this function to a variable and work with it instead of calling this function multiple times.

Why not simply calling it `getCurrentUserId()`? In difference to other places in the code where one can decide to use the cached or uncached variant, ther eis no choice here, since the other variant is private. Also this `FromRequest` from my point of view doesn't add more information, especially since there aren't any other ways for fetching the user id.

Reply:
> I must confess: Especially because of the lengths of the function name I will always assign the return value of this function to a variable and work with it instead of calling this function multiple times.

This is entirely fine, both ways are valid ways to interact with it.

However, the previous way of directly calling the `self::$current_user_id` wasn't fine, because it needed to be initialized first. This is why we needed a method instead. `getCurrentUserId` simply isn't specific enough as a name for me, because it requires additional context to know what it does, but if it really is a problem for you to use more verbose method name, we can add a proxy method.

Reply:
I do understand the argument that fetching the class variables directly without getter or setter functions is not so nice. Only thing is: I want to keep function names as brief as possible without losing essential information and I want to use  longer names only when we need to distinguish similar functions, like `DBA::select()`, `DBA::selectFirst()` or `DBA::selectToArray()`. We don't need to have called it `DBA::selectFirstToArray()` since there is only one function in the `select` family that returns only the first row and we can see from the documentation that it returns an array.

Reply:
What's the advantage of briefness given the additional context that is required to know to use short-named methods?

Reply:
The shorter a function name, the less you have to type, the more code can fit on the screen. Also it improves the readability (at least for me). I don't need to know the whole background story. For example: When we would sat together at breakfast and there would be a single piece of cheese on the table, then I would say: "Can you give me the cheese?" And not "can you give me the cheddar from the Irish company Kerrygold that riped for 6 months and that you bought last Friday at 9:00 from Trader Joe's." When there is only a single cheese we can be much more brief.

Also I'm northern German. We only use as much words as needed. :grin:

Reply:
I agree with you on principle about the more code on screen the better, but in practice we are far from it, and thankfully. There are some dreadful examples of hyper-condensed functional code that is completely unreadable, and ultimately this is the goal of code we write: to be clear and intuitive to humans. Briefness implies a lot of context that just can’t fit in short names.

To go back to your example, “Can you give me **the cheese**?” implies that both you and the person you are asking have the same knowledge of what’s on the breakfast table and what isn’t, which, even in real life, isn’t likely. What if your vision is obstructed by a milk carton and you can’t see there are two cheeses but they can? Then they have to ask you a follow-up question “Which one?”, to which you will answer slightly irritated “This one!” and point to it, to which they will reply “Ah, this one!” and finally pass it to you. A more detailed question would have lifted any potential confusion preemptively without costing you that much more in the first place.

To return to our code example, the `Cached` detail is important because the answer will depend on when the value was retrieved first, and so there is always a chance that it will be stale. What about an API endpoint that refreshes the Token? What about an impersonation endpoint that would change the current user ID? In both cases the cached method would return stale data and a call to the original method would be needed to ensure data consistency. But this important nuance would he lost with a single method with a generic short name.

Reply:
When you see the danger of using stale data, shouldn't we then avoid using cached data anyway? Thing is: I'm confident that there is no measurable time benefit between using cached data in this case in difference to fetching it from a really small database table.

Reply:
I believe the danger can be at least partly mitigated with meaningful names. But this isn’t the point of this PR. The current user id was cached in the class variable before my refactoring and I kept this behavior intact.

Reply:
I still think that the "cached" is superfluous here, since the caching is only done per request. Also we simply don't have a measurable advantage, We have additional functions and longer function names instead - but no real performance advantage.

BTW: Sometimes I see the positive effect that the language I'm using at work does limit function names to 30 characters 😁

Reply:
Here’s the thing though: If I hadn’t renamed this method explicitly, we would never have had a discussion on whether caching is relevant in this case, even though we were already using a per request cache before.

I don’t mind removing the Cached method, since this time I know it will be meaningful.

Reply:
The previous way was a "working but not so elegant" way, aka my personal coding style. (As you know I tend to work in iterations, means: At first it must work, then in the enxt steps the code itself will be improved)

Whatever we do: I have a problem with longer function names, they seem to blur in my mind and reduce the readability for me. (just like commands that are stretched to several lines. They let my brain lose the focus - and yes, I have a weird brain, I guess)

Reply:
It isn’t ideal. While you are the main contributor to this project and losing you would signify its probable death by lack of maintenance, I still believe we should make the code as accessible as possible without requiring extensive implicit knowledge, and slightly longer more meaningful names are a simple way to contribute to that goal. I don’t have a magical solutions to this, shorter names simply carry less information, there’s no way around that, and since you know most of Friendica, you have no personal use for explicit information.

Reply:
I am most likely not neurotypical, which doesn't make stuff easier 😁 

Good thing is: Most stuff at the API is done. Hopefully now most work is just bug fixing. So we can see how good this can be done.

Reply:
That I believe, and I probably am not either. And I don’t mind taking the time you need to come around specific programming concepts. Making something accessible for someone specific usually improves the overall accessibility for everyone else, but keeping method names cruelly short feels like it would penalize everybody but yourself.